kubeadm deployment instructions for Windows (#15750)
kubernetes/website: content/en/docs/setup/production-environment/windows/user-guide-windows-nodes.md

        }
    ```

> 1. Apply the Flannel yaml and validate

Reviewer: How about this optional tweak:
```suggestion
1. Apply the Flannel manifest and validate
```

  * Enable Windows Server containers role (and reboot)
  * Download and install the chosen container runtime
  * Download Kubernetes binaries and add them to the $PATH environment variable
  * Download CNI plugins based on the selection made in the Kubernetes Configuration file
  * (Optionally) Generate a new SSH key which is required to connect to the control-plane ("Master") node during joining
      {{< note >}}
      For the final step where a new ssh key is generated, the public key also needs to be added to the `authorized_keys` list file on the Linux control-plane node. This action only needs to be performed once. Steps for this are printed by the script itself at the end.  
      {{< /note >}}

#### Join the Windows Node to the Kubernetes cluster
In this section we'll cover how to join a [Windows Node with Kubernetes installed](#preparing-a-windows-node) with an existing Linux control-plane ("Master") node to form a cluster.

Use the previously downloaded [KubeCluster.ps1](https://github.com/kubernetes-sigs/sig-windows-tools/blob/master/kubeadm/v1.15.0/KubeCluster.ps1) script to join the Windows node to the cluster:

    ```PowerShell

>     .\KubeCluster.ps1 -ConfigFile .\Kubeclustervxlan.json -join 

Reviewer 2: does this doc mention the current state of vxlan vs overlay and the fact that vxlan requires a newer version of Windows?

Reviewer 2: should we recommend overlay as the default?

Author: Hmmm... My thought was that l2bridge needs additional work creating UDRs on most public cloud platforms, overlay should work anywhere out of the box. We should show simplest path to getting Kubernetes working as default. The previous docs also showed overlay as default. What are your thoughts for l2bridge as default?

But yes, I will show that overlay requires v1.14 at least and WS2019!

Reviewer 2: yeah, i think we should outline some of the trade-offs we have for the methods, currently.

Author: Sure, this item should be in https://kubernetes.io/docs/setup/production-environment/windows/intro-windows-in-kubernetes/#networking though or is something missing?

Reviewer 2: the table in there is good. i forgot about it.
ideally it should have a minimal supported Windows Version column too.

Author: Thanks for the suggestion! But in https://kubernetes.io/docs/setup/production-environment/windows/intro-windows-in-kubernetes/#windows-containers-in-kubernetes it states that:
"Windows Server 2019 is the only Windows operating system supported"

So since we require Windows Server 2019 for Kubernetes overall irrespective of networking choice, wouldn't that make the column superfluous?

Author: I added a note under the deployment instructions directly as well to highlight this though:
https://github.com/kubernetes/website/pull/15750/commits/e3d9cb40e1bdf9c39f9474eda7e68a9ee1e666f6#diff-f8c172f80db498dfcda26f3813cfa729R252

Reviewer 2: there are specific versions of Windows Server 2019 that does not support vxlan.
i encountered this issue when testing kubeadm worker nodes.

Author: OK, in this case, I will add a note in the table again that you need KB4489899 installed. The note in the deployment states this already as well, thanks!

Author: added in https://github.com/kubernetes/website/pull/15750/files#diff-b563bff111c0a208abb2db63f1e2bad4R136

    ```
    where `-ConfigFile` points to the path of the Kubernetes configuration file.

![alt_text](../kubecluster.ps1-join.gif "KubeCluster.ps1 join output")

{{< note >}}
 Should the script fail during the bootstrap or joining procedure for whatever reason, consecutive join attempts should be done in a fresh Powershell window.  
{{< /note >}}

This step will perform the following actions:
  * Establish SSH session with the Kubernetes control-plane ("Master") node to retrieve the [Kubeconfig file](https://kubernetes.io/docs/concepts/configuration/organize-cluster-access-kubeconfig/) for cluster access 
  * Register Kubelet as a Windows service
  * Configure CNI network plugins
  * Create a HNS network on top of the chosen network interface
      {{< note >}}
      This may cause a network blip for a few seconds while the vSwitch is being created.
      {{< /note >}}
  * (If vxlan plugin is selected) Open up inbound firewall UDP port 4789 for overlay traffic
  * Register FlannelD as a Windows service

>   * Register Kube-proxy as a Windows service

Reviewer 2: some of the component names in the above list probably should be lowercased as in the rest of the k8s.io documentation - e.g. kube-proxy, kubelet.

Author: addressed in the newest commit, thank you!

All code snippets in Windows sections are to be run in a PowerShell environment with elevated permissions (Admin).
{{< /note >}}

1. Pull down the [SIG Windows tools](https://github.com/kubernetes-sigs/sig-windows-tools) repository 
   ```PowerShell

>    git clone https://github.com/kubernetes-sigs/sig-windows-tools.git

Reviewer 2: /cc @ksubrmnn

    kube-proxy        2         2         2       2            2           beta.kubernetes.io/os=linux                                 26d
    ```

#### Join Windows Worker

In this section we'll cover configuring a Windows node from scratch to join a cluster on-prem. If your cluster is on a cloud you'll likely want to follow the cloud specific guides in the next section.

#### Preparing a Windows Node

{{< note >}}
All code snippets in Windows sections are to be run in a PowerShell environment with elevated permissions (Administrator) on the Windows worker node.
{{< /note >}}

1. Download the [SIG Windows tools](https://github.com/kubernetes-sigs/sig-windows-tools) repository containing install and join scripts
   ```PowerShell
   [Net.ServicePointManager]::SecurityProtocol = [Net.SecurityProtocolType]::Tls12
   Start-BitsTransfer https://github.com/kubernetes-sigs/sig-windows-tools/archive/master.zip
   tar -xvf .\master.zip --strip-components 3 sig-windows-tools-master/kubeadm/v1.15.0/*
   Remove-Item .\master.zip
   ```

1. Customize the Kubernetes [configuration file](https://github.com/kubernetes-sigs/sig-windows-tools/blob/master/kubeadm/v1.15.0/Kubeclustervxlan.json) 

>     ```

Reviewer: Suggestion: use a `codenew` Hugo shortcode, and place this in a separate file

Author: Thanks for the suggestion - This is showing up as a pointer to the old instructions, which chunk of code are you referring to?

        },
        "Network" : {  // Contains values for IP ranges in CIDR notation for Kubernetes networking
            "ServiceCidr" : "10.96.0.0/12",  // Service IP subnet used by Services in CIDR notation
            "ClusterCidr" : "10.244.0.0/16"  // Cluster IP subnet used by Pods in CIDR notation 
        }
    },
    "Install" : {  // Contains values and configurations for Windows Node installation
        "Destination" : "C:\\ProgramData\\Kubernetes"  // Absolute DOS path where Kubernetes will be installed on the Windows Node
    }
}
  ```

1. Install containers and Kubernetes (requires a system reboot)

Use the previously downloaded [KubeCluster.ps1](https://github.com/kubernetes-sigs/sig-windows-tools/blob/master/kubeadm/v1.15.0/KubeCluster.ps1) script to install Kubernetes on the Windows Server container host:

  ```PowerShell
    .\KubeCluster.ps1 -ConfigFile Kubeclustervxlan.json -install 
  ```
  where `-ConfigFile` points to the path of the Kubernetes configuration file.

  ![alt_text](../kubecluster.ps1-install.gif "KubeCluster.ps1 install output")

> This step will perform the following actions on a targeted Windows node:

Reviewer: ```suggestion
On the Windows node you target, this step will:
```

  * Enable Windows Server containers role (and reboot)
  * Download and install the chosen container runtime

>   * Download Kubernetes binaries and add them to the $PATH environment variable

Reviewer: ```suggestion
  * Download Kubernetes binaries and add them to the `$PATH` environment variable
```

  * Download CNI plugins based on the selection made in the Kubernetes Configuration file

>   * (Optionally) Generate a new SSH key which is required to connect to the control-plane ("Master") node during joining

Reviewer: ```suggestion
  * (Optionally) Generate a new SSH key pair. The SSH key lets you connect to the control plane during joining
```

      {{< note >}}

>       For the final step where a new ssh key is generated, the public key also needs to be added to the `authorized_keys` list file on the Linux control-plane node. This action only needs to be performed once. Steps for this are printed by the script itself at the end.  

Reviewer: ```suggestion
      For the SSH key generation step, you also need to add the generated public SSH key to the `authorized_keys` file on your (Linux) control plane node. You only need to do this once. The script prints out the steps you can follow to do this, at the end of its output.
```

      {{< /note >}}

    ```PowerShell
    [Environment]::SetEnvironmentVariable("HTTP_PROXY", "http://proxy.example.com:80/", [EnvironmentVariableTarget]::Machine)
    [Environment]::SetEnvironmentVariable("HTTPS_PROXY", "http://proxy.example.com:443/", [EnvironmentVariableTarget]::Machine)
    ```

#### Join the Windows Node to the Kubernetes cluster

> In this section we'll cover how to join a [Windows Node with Kubernetes installed](#preparing-a-windows-node) with an existing Linux control-plane ("Master") node to form a cluster.

Reviewer: ```suggestion
This section covers how to join a [Windows Node with Kubernetes installed](#preparing-a-windows-node) with an existing (Linux) control plane, to form a cluster.
```

Use the previously downloaded [KubeCluster.ps1](https://github.com/kubernetes-sigs/sig-windows-tools/blob/master/kubeadm/v1.15.0/KubeCluster.ps1) script to join the Windows node to the cluster:

    ```PowerShell
    .\KubeCluster.ps1 -ConfigFile .\Kubeclustervxlan.json -join 
    ```
    where `-ConfigFile` points to the path of the Kubernetes configuration file.

![alt_text](../kubecluster.ps1-join.gif "KubeCluster.ps1 join output")

{{< note >}}

>  Should the script fail during the bootstrap or joining procedure for whatever reason, consecutive join attempts should be done in a fresh Powershell window.  

Reviewer: ```suggestion
Should the script fail during the bootstrap or joining procedure for whatever reason, start a new PowerShell session before starting each consecutive join attempt.
```
(the reader might plausibly not be using PowerShell in a window).

{{< /note >}}

This step will perform the following actions:

>   * Establish SSH session with the Kubernetes control-plane ("Master") node to retrieve the [Kubeconfig file](https://kubernetes.io/docs/concepts/configuration/organize-cluster-access-kubeconfig/) for cluster access 

Reviewer: ```suggestion
  * Connect to the control plane ("Master") host via SSH, to retrieve the [Kubeconfig file](/docs/concepts/configuration/organize-cluster-access-kubeconfig/) file.
```

            "Name" : "flanneld",  // Name of network daemon
            "Url" : "https://github.com/coreos/flannel/releases/download/v0.11.0/flanneld.exe"  // Direct URL pointing to network daemon executable
            }
        ],
        "Plugin" : {  // Contains values for CNI network plugin
            "Name": "vxlan" // Backend network mechanism to use: ["vxlan" | "bridge"]
        },
        "InterfaceName" : "Ethernet"  // Designated network interface name on Windows node to use as container network
    },
    "Kubernetes" : {  // Contains values for Kubernetes node binaries
        "Source" : {  // Contains values for Kubernetes node binaries 
            "Release" : "1.15.0",  // Version of Kubernetes node binaries
            "Url" : "https://dl.k8s.io/v1.15.0/kubernetes-node-windows-amd64.tar.gz"  // Direct URL pointing to Kubernetes node binaries tarball
        },
        "ControlPlane" : {  // Contains values associated with Kubernetes control-plane ("Master") node
            "IpAddress" : "kubemasterIP",  // IP address of control-plane ("Master") node
            "Username" : "localadmin",  // Username on control-plane ("Master") node with remote SSH access  

>             "KubeadmToken" : "token",  // Kubeadm bootstrap token

Reviewer 3: Include that these values can be generated via `kubeadm token create --print-join-command` on the control plane node

Author: Included as a note

{{< /note >}}

  ![alt_text](../kubecluster.ps1-install.gif "KubeCluster.ps1 install output")

On the Windows node you target, this step will:
  1. Enable Windows Server containers role (and reboot)
  1. Download and install the chosen container runtime
  1. Download all needed container images
  1. Download Kubernetes binaries and add them to the `$PATH` environment variable
  1. Download CNI plugins based on the selection made in the Kubernetes Configuration file
  1. (Optionally) Generate a new SSH key which is required to connect to the control-plane ("Master") node during joining
      {{< note >}}
      For the SSH key generation step, you also need to add the generated public SSH key to the `authorized_keys` file on your (Linux) control-plane node. You only need to do this once. The script prints out the steps you can follow to do this, at the end of its output.
      {{< /note >}}

> #### Join the Windows Node to the Kubernetes cluster

Reviewer 3: Before this, can we include a section that says that if you want to replace binaries or edit configs, you should do that post install and pre join

Author: Added this

This section covers how to join a [Windows node with Kubernetes installed](#preparing-a-windows-node) with an existing (Linux) control-plane, to form a cluster.

Use the previously downloaded [KubeCluster.ps1](https://github.com/kubernetes-sigs/sig-windows-tools/blob/master/kubeadm/v1.15.0/KubeCluster.ps1) script to join the Windows node to the cluster:

  ```PowerShell
    .\KubeCluster.ps1 -ConfigFile .\Kubeclustervxlan.json -join 
  ```
  where `-ConfigFile` points to the path of the Kubernetes configuration file.

![alt_text](../kubecluster.ps1-join.gif "KubeCluster.ps1 join output")

{{< note >}}
Should the script fail during the bootstrap or joining procedure for whatever reason, start a new PowerShell session before starting each consecutive join attempt.
{{< /note >}}

This step will perform the following actions:
  1. Connect to the control-plane ("Master") node via SSH, to retrieve the [Kubeconfig file](/docs/concepts/configuration/organize-cluster-access-kubeconfig/) file.
  1. Register kubelet as a Windows service
  1. Configure CNI network plugins

>   1. Create a HNS network on top of the chosen network interface

Reviewer 3: *an HNS network

Author: corrected now :)

      {{< note >}}
      This may cause a network blip for a few seconds while the vSwitch is being created.
      {{< /note >}}
  1. (If vxlan plugin is selected) Open up inbound firewall UDP port 4789 for overlay traffic
  1. Register flanneld as a Windows service
  1. Register kube-proxy as a Windows service

Now you can view the Windows nodes in your cluster by running the following:

```bash
kubectl get nodes
```

#### Remove the Windows Node from the Kubernetes cluster
In this section we'll cover how to remove a Windows node from a Kubernetes cluster. 

Use the previously downloaded [KubeCluster.ps1](https://github.com/kubernetes-sigs/sig-windows-tools/blob/master/kubeadm/v1.15.0/KubeCluster.ps1) script to remove the Windows node from the cluster:

  ```PowerShell
    .\KubeCluster.ps1 -ConfigFile .\Kubeclustervxlan.json -reset 
  ```
  where `-ConfigFile` points to the path of the Kubernetes configuration file.

![alt_text](../kubecluster.ps1-reset.gif "KubeCluster.ps1 reset output")

This step will perform the following actions on the targeted Windows node:
  1. Delete the Windows node from the Kubernetes cluster

>   1. Stop all running containers and remove all container images

Reviewer 3: I don't believe we remove all container images

Author: Hm not sure why I was seeing that but I updated this to just say "Stop all running containers" now (I verified in the script that we still do that by running `docker rm` on all the containers). Thanks!

    The "pause" (infrastructure) image is hosted on Microsoft Container Registry (MCR). You can access it using "docker pull mcr.microsoft.com/k8s/core/pause:1.2.0". The DOCKERFILE is available at https://github.com/kubernetes-sigs/sig-windows-tools/tree/master/cmd/wincat.
    {{< /note >}}

> 3. Prepare a Windows directory for Kubernetes

Reviewer 2: i think the numbering should be `1.` for all entries actually, because there are problems with the markdown parser.

#### Deployment with kubeadm and cluster API

Kubeadm is becoming the de facto standard for users to deploy a Kubernetes cluster. Windows node support in kubeadm is an alpha feature since Kubernetes release v1.16. We are also making investments in cluster API to ensure Windows nodes are properly provisioned. For more details, please consult the [kubeadm for Windows KEP](https://github.com/kubernetes/enhancements/blob/master/keps/sig-cluster-lifecycle/kubeadm/20190424-kubeadm-for-windows.md).

### Next Steps

Now that you've configured a Windows worker in your cluster to run Windows containers you may want to add one or more Linux nodes as well to run Linux containers. You are now ready to schedule Windows containers on your cluster.

> ### FAQs

Reviewer: ```suggestion
### Troubleshooting
```

#### My Kubernetes installation is failing because my Windows Server node is behind a proxy 
If you are behind a proxy, the following PowerShell environment variables must be defined:
  ```PowerShell
  [Environment]::SetEnvironmentVariable("HTTP_PROXY", "http://proxy.example.com:80/", [EnvironmentVariableTarget]::Machine)
  [Environment]::SetEnvironmentVariable("HTTPS_PROXY", "http://proxy.example.com:443/", [EnvironmentVariableTarget]::Machine)
  ```

> #### What is the "pause" container image?

Reviewer: ```suggestion
### `pause` container
```
